### PR TITLE
Hocs foi

### PR DESCRIPTION
**Pull request by CPage-Ten10**

### CODE REVIEW

File: src/test/java/com/hocs/test/pages/foi/SoftClose.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.hocs.test.pages.foi;
+
+import com.hocs.test.pages.decs.BasePage;
+
+public class SoftClose extends BasePage {
+
```
Comment (CaseyProsser) [inline src/test/java/com/hocs/test/pages/foi/SoftClose.java:6]: same as above, are we missing something here or is this a placeholder for a later development?

File: src/test/java/com/hocs/test/pages/ukvi/MPAMMultipleContributions.java [hunk 2]
```diff
@@ -51,8 +52,14 @@ public void sendMPAMCaseToContributionRequest() {
     }
 
     public void addMultipleContributionRequests(int numberOfContributions) {
```
Comment (CaseyProsser) [inline src/test/java/com/hocs/test/pages/ukvi/MPAMMultipleContributions.java:54]: We need to refactor this so that it makes sense for both MPAM and FOI. We should either create a separate file for FOI, or move this method to a file thats in the CS folder.

File: src/test/resources/features/comp/COMPSearch.feature [hunk 3]
```diff
@@ -25,7 +25,7 @@ Feature: COMP Search
   @COMPRegression
   Scenario: User can search for a COMP case by its case reference
     When I create a single "COMP" case
-    And I search for the COMP case by its case reference
+    And I search for the "COMP" case by its case reference
```
Comment (CaseyProsser) [inline src/test/resources/features/comp/COMPSearch.feature:28]: seeing as this doesnt ues the casetype variable, maybe it should just read 'And I search for the case by its case reference'

File: src/test/resources/features/foi/Acceptance.feature [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+  Background:
+    Given I am logged into "CS" as user "FOI_USER"
+    When I create a "FOI" case and move it to the "Acceptance" stage
+    And I load and claim the current case
+
+  #HOCS-2586, HOCS-2723
+  Scenario: User is able to complete the Acceptance stage
+    And I select that the case "Does" belong to this Directorate
+    And I click the "Continue" button
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/Acceptance.feature:12]: can we move the button clicks like this into the steps above them?

File: src/test/resources/features/foi/Acceptance.feature [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    And the case should be assigned to the Drafting team selected at Acceptance
+
+  #HOCS-2753, HOCS-2741
+  @FOIRegression
+  Scenario: User rejects a case at the Acceptance stage
+    And I select that the case "Doesn't" belong to this Directorate
+    And I enter a rejection reason at the Acceptance stage
+    And I click the "Continue" button
+    Then the case should be returned to the "Allocation" stage
+    And a rejection note should be visible showing the reason for rejection
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/Acceptance.feature:25]: do foi workstacks have a 'Rejected' column? 
If so we should include the standard assertion that the column is populated with the right value

File: src/test/resources/features/foi/Allocation.feature [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+  Background:
+    Given I am logged into "CS" as user "FOI_USER"
+    When I create a "FOI" case and move it to the "Allocation" stage
+    And I load and claim the current case
+
+  #HOCS-3626 #HOCS-3268 #HOCS-2326
+  @FOIRegression
+  Scenario: User is able to complete the Allocation stage
+    And the Requested Question should be displayed in the summary tab
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/Allocation.feature:12]: why is this assertion at the start of this scenario?
should it be its own scenario? or is the step just worded so it doesnt make sense in this script (maybe reword it?)

File: src/test/resources/features/foi/Approval.feature [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  Background:
+    Given I am logged into "CS" as user "FOI_USER"
+    When I create a "FOI" case and move it to the "Approval" stage
+    And I load and claim the current case
+
+  Scenario: User is able to complete the Approval stage
+    And I add an Approval request to the case with the "Complete" status
+    And I click the "Continue" button
+    Then the case should be moved to the "Dispatch" stage
+
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/Approval.feature:13]: why no assertion on the team name?

File: src/test/resources/features/foi/CaseCreationStage.feature [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  Background:
+    Given I am logged into "CS" as user "FOI_USER"
+    When I create a "FOI" case and move it to the "Case Creation" stage
+    And I load the current case
+
+  #HOCS-3408
+  Scenario: User is able to complete the Case Creation stage
+    And I click the "Continue" button
+    And I select "Yes" for the validity of the request and continue
+    And I click the "Continue" button
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/CaseCreationStage.feature:13]: All 3 of these steps could be summed up in one step like 'And I submit that this is a valid request' or something. We want to take as many of the 'I click x' steps out as possible from now on

File: src/test/resources/features/foi/ConsiderAndDraft.feature [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    And I select that the case "Doesn't" require a contribution
+    And I click the "Continue" button
+    And I select "Full disclosure" as the response type
+    And I click the "Continue" button
+    And I click add documents
+    And I choose the document type "Draft response"
+    And I upload a file of type "docx"
+    And I click the "Continue" button
+    Then the case should be moved to the "Approval" stage
+
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/ConsiderAndDraft.feature:21]: For this we should move the click buttons into the steps above, maybe rename them so they are move descriptive of the users goal than the specific action on the application, and the 3 document steps should all be one thats like 'I add a Draft Response document'

File: src/test/resources/features/foi/FOICreateCase.feature [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+  #HOCS-3251, HOCS-3248, HOCS-3247, HOCS-3252
+  Scenario: User is able to create an FOI Case
+    And I select how the request was received
+    And I enter the correspondent details
+    And I select the "Animal alternatives (3Rs)" FOI topic
+    And I enter the Request Question
+    And I click the "Submit" button
+    Then a case is created successfully
+    And I navigate to the "Dashboard" page
+    Then the case should be moved to the "Case Creation" stage
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/FOICreateCase.feature:18]: this assertion step has a 'gotodashbaord()' call in it, so we dont need the navigate step above

File: src/test/resources/features/foi/FOIDispatch.feature [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+  Scenario: User is able to complete the Dispatch stage
+    And I select "Yes" to do you want to dispatch the case
+    And I click the "Continue" button
+    Then the case should be moved to the "Soft Close" stage
+
+  @FOIRegression
+  Scenario: User is able to reject a case at the Dispatch stage
+    And I select "No" to do you want to dispatch the case
+    And I click the "Continue" button
+    Then the case should be moved to the "Approval" stage
```
Comment (CaseyProsser) [inline src/test/resources/features/foi/FOIDispatch.feature:18]: no rejection note for this scenario?

File: src/test/java/com/hocs/test/glue/foi/SoftCloseStepDefs.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package com.hocs.test.glue.foi;
+
+import com.hocs.test.pages.decs.BasePage;
+
+public class SoftCloseStepDefs extends BasePage {
+
```
[the remark this answers is not in the document]
Comment (CPage-Ten10) [reply src/test/java/com/hocs/test/glue/foi/SoftCloseStepDefs.java:6]: Yeah this is more of a placeholder since the only thing to do at Soft Close is to reopen the case and I've used the 'And I click the "Reopen" button' step instead of writing up a different one

File: src/test/java/com/hocs/test/pages/foi/SoftClose.java (hunk 1 above)
Comment (CPage-Ten10) [reply src/test/java/com/hocs/test/pages/foi/SoftClose.java:6]: See other comment

File: src/test/java/com/hocs/test/pages/ukvi/MPAMMultipleContributions.java (hunk 2 above)
Comment (CPage-Ten10) [reply src/test/java/com/hocs/test/pages/ukvi/MPAMMultipleContributions.java:54]: See comment regarding Approval requests

File: src/test/resources/features/foi/Acceptance.feature (hunk 4 above)
Comment (CPage-Ten10) [reply src/test/resources/features/foi/Acceptance.feature:12]: Resolved

File: src/test/resources/features/comp/COMPSearch.feature (hunk 3 above)
Comment (CPage-Ten10) [reply src/test/resources/features/comp/COMPSearch.feature]: Resolved

File: src/test/java/com/hocs/test/glue/foi/SoftCloseStepDefs.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (CaseyProsser) [reply src/test/java/com/hocs/test/glue/foi/SoftCloseStepDefs.java:6]: understood

File: src/test/java/com/hocs/test/pages/decs/Search.java [hunk 13]
```diff
@@ -595,6 +632,76 @@ public void assertCOMPInformationRandomSearchResult(String criteria) {
...
+                    assertThat(trueFalse, is(true));
+                    n++;
+                }
+                break;
+            case "CORRESPONDENT (NON-MP)":
+                String correspondentName = sessionVariableCalled("searchCorrespondentName").toString().toUpperCase();
+                List<WebElementFacade> listOfCasesWithCorrespondent = findAll("//td[contains(text(), '" + correspondentName + "')]");
+                while (n < listOfCasesWithCorrespondent.size()) {
+                    assertThat(listOfCasesWithCorrespondent.get(n).getText().toUpperCase().contains(correspondentName), is(true));
+                    n++;
```
[the remark this answers is not in the document]
Comment (CaseyProsser) [reply src/test/java/com/hocs/test/pages/decs/Search.java:687]: understood, just not sure of the logic here. We are getting a set of elements that contain the name, and then checking they contain the name? Am I misreading it?

File: src/test/java/com/hocs/test/pages/decs/Search.java [hunk 14]
```diff
@@ -595,6 +632,76 @@ public void assertCOMPInformationRandomSearchResult(String criteria) {
...
+                break;
+            case "TOPIC":
+                safeClickOn(randomSearchResultHypertext);
+                summaryTab.selectSummaryTab();
+                String displayedTopic = summaryTab.foiTopic.getText();
+                String searchedTopic = sessionVariableCalled("searchTopic");
+                assertThat(displayedTopic.equals(searchedTopic), is(true));
+                break;
+            case "ACTIVE CASES ONLY":
+                //This doesn't really work since the 'soft closed' cases are still technically active
```
[the remark this answers is not in the document]
Comment (CaseyProsser) [reply src/test/java/com/hocs/test/pages/decs/Search.java:698]: I agree, thats not a good option for this scenario. Ill close this point

File: src/test/java/com/hocs/test/pages/foi/Approval.java [hunk 15]
```diff
@@ -0,0 +1,92 @@
+package com.hocs.test.pages.foi;
```
[the remark this answers is not in the document]
Comment (CaseyProsser) [reply src/test/java/com/hocs/test/pages/foi/Approval.java:1]: Yeah. Ill take a look at that separately I guess, doesnt have to be part of this FOI work.

File: src/test/resources/features/foi/Allocation.feature (hunk 6 above)
Comment (CaseyProsser) [reply src/test/resources/features/foi/Allocation.feature:11]: Yeh I think this should be 2 scenarios. The assertion at the start has nothing to do with the scenario title

File: src/test/resources/serenity.conf [hunk 16]
```diff
@@ -21,7 +21,7 @@ zalenium {
   }
 environments {
     default {
-      login.page.url = "https://qa.internal.cs-notprod.homeoffice.gov.uk/"
+      login.page.url = "https://delta.internal.cs-notprod.homeoffice.gov.uk/"
```
Comment (CaseyProsser) [inline src/test/resources/serenity.conf:24]: this needs changing back to qa

File: src/test/resources/features/foi/Acceptance.feature (hunk 4 above)
Comment (CPage-Ten10) [reply src/test/resources/features/foi/Acceptance.feature:12]: In this instance it wouldn't work, the scenario directly after this has the step 'And I enter a rejection reason at the Acceptance stage', this must be completed before the user clicks the continue button.

File: src/test/resources/features/foi/CaseCreationStage.feature (hunk 8 above)
Comment (CPage-Ten10) [reply src/test/resources/features/foi/CaseCreationStage.feature:13]: The selecting validity step already has a click continue button in it, the one directly following it is for the screen after the validity selection screen

File: src/test/resources/features/foi/FOIDispatch.feature (hunk 11 above)
Comment (CPage-Ten10) [reply src/test/resources/features/foi/FOIDispatch.feature:19]: I think I'll have to raise a bug, a rejection note isn't produced when the case is sent back

**Review by CaseyProsser — APPROVED** [review]
